radicle-cob/src/sign.rs

```diff
@@ -0,0 +1,45 @@
+// Copyright © 2022 The Radicle Link Contributors
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+use std::error::Error;
+
+pub mod message;
+pub use message::CommitMessage;
+
+pub mod signatures;
+pub use signatures::{Signature, Signatures};
+
+pub trait Signer {
+    type Error: Error + Send + Sync + 'static;
+    type Signature: Clone + ToString + AsRef<[u8]>;
+    type Key: PublicKey<Signature = Self::Signature> + Clone + ToString;
+
+    fn sign(&self, bytes: &[u8]) -> Result<Self::Signature, Self::Error>;
+    fn public_key(&self) -> Self::Key;
+}
+
+pub trait PublicKey {
+    type Signature: ToString + AsRef<[u8]>;
+
+    fn verify(&self, signature: &Self::Signature, bytes: &[u8]) -> bool;
+}
+
+pub trait Key
+where
+    Self: PublicKey + ToString + Ord,
+    Self::Signature: ToString,
+    for<'de> Self: serde::Deserialize<'de>,
+    for<'de> Self::Signature: serde::Deserialize<'de>,
+{
+}
+
+impl<T> Key for T
+where
+    Self: PublicKey + ToString + Ord,
+    Self::Signature: ToString,
+    for<'de> Self: serde::Deserialize<'de>,
+    for<'de> Self::Signature: serde::Deserialize<'de>,
+{
+}
```

Comment:
I wonder if it makes sense to have a `radicle-crypto` crate, that `radicle-cob` can depend on, so that we don't need generic key types. It would simplify things a lot.

Reply:
> I wonder if it makes sense to have a `radicle-crypto` crate, that `radicle-cob` can depend on, so that we don't need generic key types. It would simplify things a lot.

My guy... I was asking this here https://discord.com/channels/841318878125490186/880000282388475925/1031513523588706355 :smile:

Reply:
Aha I see, yes, the answer is yes :)